docs/syntax/Cadence.syntax

```diff
@@ -0,0 +1,1482 @@
+    |   destroyFunctionDeclaration
+    |   prepareFunctionDeclaration
+    |   functionDeclaration
+    |   structCompositeDeclaration
+    |   resourceCompositeDeclaration
+    |   contractCompositeDeclaration
+    |   enumCompositeDeclaration
+    |   structInterfaceDeclaration
+    |   resourceInterfaceDeclaration
+    |   contractInterfaceDeclaration
+    |   enumInterfaceDeclaration
+    |   eventDeclaration
+    |   pragmaDeclaration
+    |   entitlementDeclaration
+    |   entitlementMappingDeclaration
+    |   attachmentDeclaration
+    ;
+
+initFunctionDeclaration:
+        access *1viewKeyword initKeyword parameterList *1functionBlock
+    ;
+
+destroyFunctionDeclaration:
+        access *1viewKeyword destroyKeyword parameterList *1functionBlock
+    ;
+
+prepareFunctionDeclaration:
+        access *1viewKeyword prepareKeyword parameterList *1functionBlock
+    ;
+
+functionDeclaration:
+        access *1viewKeyword funKeyword nonReservedIdentifier parameterList *1functionDeclarationTypeAnnotation *1functionBlock
+    ;
+
+functionDeclarationTypeAnnotation:
+        ":" typeAnnotation
+    ;
+
+eventDeclaration:
+        access eventKeyword nonReservedIdentifier parameterList
+    ;
+
+comma:
+        ","
+    ;
+
+pragmaDeclaration:
+        "#" expression
+    ;
+
+entitlementDeclaration:
+        access entitlementKeyword nonReservedIdentifier
+    ;
+
+entitlementMappingDeclaration:
+        access entitlementKeyword mappingKeyword nonReservedIdentifier "{" entitlementMappingsAndInclusions "}"
+    ;
+
+entitlementMappingsAndInclusions:
+        *entitlementMappingElement
+    ;
+
+entitlementMappingElement:
+        entitlementMapping
+    |   entitlementIncludeType
+    ;
+
+entitlementMapping:
+        entitlementMappingSource "->" entitlementMappingTarget
+    ;
+
+entitlementMappingSource:
+        fullType
+    ;
+
+entitlementMappingTarget:
+        fullType
+    ;
+
+entitlementIncludeType:
+        includeKeyword nominalType
+    ;
+
+semicolonEnumerationDeclarationCase:
+        semicolon
+    |   enumerationDeclarationCase
+    ;
+
+enumerationDeclarationCase:
+        access caseKeyword nonReservedIdentifier
+    ;
+
+parameterList:
+        "(" *parameter( "," ) *1comma ")"
+    ;
+
+parameter:
+        parseLabeledParameter
+    |   parseUnlabeledParameter
+    ;
+
+parseLabeledParameter:
+        parameterPublicName parameterLocalName ":" typeAnnotation *1defaultValue
+    ;
+
+parseUnlabeledParameter:
+        parameterLocalName ":" typeAnnotation *1defaultValue
+    ;
+
+parameterPublicName:
+        nonReservedIdentifier
+    ;
+
+parameterLocalName:
+        nonReservedIdentifier
+    ;
+
+defaultValue:
+        "=" expression
+    ;
+
+typeAnnotation:
+        *1resourceAnnotation fullType
+    ;
+
+resourceAnnotation:
+        "@"
+    ;
+
+fullType:
+        fullTypeNormal
+    |   fullTypeNested
+    ;
+
+fullTypeNormal:
+        *1fullTypeReference innerType *fullTypeOptional
+    ;
```

Comment:
Maybe encode the precedence rules for types in nested rules, like it is done for expressions, see the `typeLeftBindingPower*` constants. For example, references bind tighter than optionals